File: sdk/sdk-python/netledger_sdk/http_client.py
"""HTTP client for making API requests."""

import requests
from requests.adapters import HTTPAdapter
from typing import Any, Dict, Optional, TypeVar, Generic
from .exceptions import NetLedgerConnectionError, NetLedgerApiError


T = TypeVar('T')


class ApiResponse(Generic[T]):
    """Represents an API response."""

    def __init__(self, data: Optional[T], status_code: int, request_id: Optional[str] = None):
        self.data = data
        self.status_code = status_code
        self.request_id = request_id


class HttpClient:
    """HTTP client for making API requests."""

    def __init__(
        self,
        base_url: str,
        api_key: str,
        timeout_seconds: float = 30.0,
        tenant_id: Optional[str] = None,
        pool_connections: int = 10,
        pool_maxsize: int = 20
    ):
        """
        Initialize the HTTP client.

        Args:
            base_url: The base URL of the NetLedger server.
            api_key: The API key for authentication.
            timeout_seconds: Request timeout in seconds.
        """
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.tenant_id = tenant_id
        self.timeout = timeout_seconds
        self.session = requests.Session()
        adapter = HTTPAdapter(pool_connections=pool_connections, pool_maxsize=pool_maxsize)
        self.session.mount('http://', adapter)
        self.session.mount('https://', adapter)
        headers = {
            'Authorization': f'Bearer {api_key}',
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        }
        if tenant_id:
            headers['x-tenant-id'] = tenant_id
        self.session.headers.update(headers)

    def get(self, path: str) -> ApiResponse:
        """Make a GET request."""
        return self._request('GET', path)

    def put(self, path: str, body: Optional[Any] = None) -> ApiResponse:
        """Make a PUT request."""
        return self._request('PUT', path, body)

    def put_raw(
        self,
        path: str,
        body: Any,
        content_type: str = 'application/octet-stream',
        headers: Optional[Dict[str, str]] = None
    ) -> ApiResponse:
        """Make a PUT request with raw content."""
        return self._raw_request('PUT', path, body, content_type, headers)

    def post(self, path: str, body: Optional[Any] = None) -> ApiResponse:
        """Make a POST request."""
        return self._request('POST', path, body)

    def delete(self, path: str) -> None:
        """Make a DELETE request."""
        self._request('DELETE', path)

    def delete_with_response(self, path: str) -> ApiResponse:
        """Make a DELETE request and return a response body."""
        return self._request('DELETE', path)

    def head(self, path: str) -> bool:
        """Make a HEAD request to check if a resource exists."""
        try:
            url = f'{self.base_url}{path}'
            response = self.session.head(url, timeout=self.timeout)
            return 200 <= response.status_code < 300
        except requests.exceptions.Timeout:
            raise NetLedgerConnectionError('Request timed out')
        except requests.exceptions.RequestException as e:
            raise NetLedgerConnectionError('Failed to connect to the server', e)

    def _request(self, method: str, path: str, body: Optional[Any] = None) -> ApiResponse:
        """Make an HTTP request."""
        try:
            url = f'{self.base_url}{path}'

            kwargs: Dict[str, Any] = {'timeout': self.timeout}
            if body is not None:
                kwargs['json'] = body

            response = self.session.request(method, url, **kwargs)

            if not (200 <= response.status_code < 300):
                error_message = response.reason or 'Unknown error'
                error_details = None

                try:
                    error_data = response.json()
                    error_message = error_data.get('message', error_message)
                    error_details = error_data.get('description')
                except (ValueError, KeyError):
                    pass

                raise NetLedgerApiError(response.status_code, error_message, error_details)

            # Get request identifier from header
            request_id = response.headers.get('x-request-id')

            if not response.text or response.text.strip() == '':
                return ApiResponse(None, response.status_code, request_id)

            try:
                data = response.json()
                return ApiResponse(
                    data=data,
                    status_code=response.status_code,
                    request_id=request_id
                )
            except ValueError:
                raise NetLedgerApiError(response.status_code, 'Failed to parse response')

        except requests.exceptions.Timeout:
            raise NetLedgerConnectionError('Request timed out')
        except requests.exceptions.RequestException as e:
            raise NetLedgerConnectionError('Failed to connect to the server', e)

    def _raw_request(
        self,
        method: str,
        path: str,
        body: Any,
        content_type: str,
        headers: Optional[Dict[str, str]] = None
    ) -> ApiResponse:
        """Make an HTTP request with raw content."""
        try:
            url = f'{self.base_url}{path}'
            request_headers = {'Content-Type': content_type or 'application/octet-stream'}
            if headers:
                request_headers.update(headers)

            response = self.session.request(
                method,
                url,
                data=body,
                headers=request_headers,
                timeout=self.timeout
            )

            if not (200 <= response.status_code < 300):
                error_message = response.reason or 'Unknown error'
                error_details = None

                try:
                    error_data = response.json()
                    error_message = error_data.get('message', error_message)
                    error_details = error_data.get('description')
                except (ValueError, KeyError):
                    pass

                raise NetLedgerApiError(response.status_code, error_message, error_details)

            request_id = response.headers.get('x-request-id')
            if not response.text or response.text.strip() == '':
                return ApiResponse(None, response.status_code, request_id)

            try:
                data = response.json()
                return ApiResponse(
                    data=data,
                    status_code=response.status_code,
                    request_id=request_id
                )
            except ValueError:
                raise NetLedgerApiError(response.status_code, 'Failed to parse response')

        except requests.exceptions.Timeout:
            raise NetLedgerConnectionError('Request timed out')
        except requests.exceptions.RequestException as e:
            raise NetLedgerConnectionError('Failed to connect to the server', e)

    def close(self) -> None:
        """Close the session."""
        self.session.close()

File: sdk/sdk-js/src/http-client.ts
import * as http from 'http';
import * as https from 'https';
import { URL } from 'url';
import { ApiResponse, ErrorResponse } from './models';
import { NetLedgerConnectionError, NetLedgerApiError } from './errors';

/**
 * HTTP client for making API requests.
 */
export class HttpClient {
    private readonly baseUrl: string;
    private readonly apiKey: string;
    private readonly timeoutMs: number;

    constructor(baseUrl: string, apiKey: string, timeoutMs: number = 30000) {
        this.baseUrl = baseUrl.replace(/\/$/, '');
        this.apiKey = apiKey;
        this.timeoutMs = timeoutMs;
    }

    /**
     * Make a GET request.
     */
    async get<T>(path: string): Promise<ApiResponse<T>> {
        return this.request<T>('GET', path);
    }

    /**
     * Make a PUT request.
     */
    async put<T>(path: string, body?: unknown): Promise<ApiResponse<T>> {
        return this.request<T>('PUT', path, body);
    }

    /**
     * Make a POST request.
     */
    async post<T>(path: string, body?: unknown): Promise<ApiResponse<T>> {
        return this.request<T>('POST', path, body);
    }

    /**
     * Make a DELETE request.
     */
    async delete(path: string): Promise<void> {
        await this.request<unknown>('DELETE', path);
    }

    /**
     * Make a HEAD request.
     */
    async head(path: string): Promise<boolean> {
        return new Promise<boolean>((resolve, reject) => {
            const url = new URL(path, this.baseUrl);
            const isHttps = url.protocol === 'https:';
            const lib = isHttps ? https : http;

            const options: http.RequestOptions = {
                method: 'HEAD',
                hostname: url.hostname,
                port: url.port || (isHttps ? 443 : 80),
                path: url.pathname + url.search,
                timeout: this.timeoutMs,
                headers: {
                    'Authorization': `Bearer ${this.apiKey}`,
                    'Accept': 'application/json'
                }
            };

            const req = lib.request(options, (res) => {
                resolve(res.statusCode !== undefined && res.statusCode >= 200 && res.statusCode < 300);
            });

            req.on('error', (err) => {
                reject(new NetLedgerConnectionError('Failed to connect to the server', err));
            });

            req.on('timeout', () => {
                req.destroy();
                reject(new NetLedgerConnectionError('Request timed out'));
            });

            req.end();
        });
    }

    /**
     * Make an HTTP request.
     */
    private request<T>(method: string, path: string, body?: unknown): Promise<ApiResponse<T>> {
        return new Promise<ApiResponse<T>>((resolve, reject) => {
            const url = new URL(path, this.baseUrl);
            const isHttps = url.protocol === 'https:';
            const lib = isHttps ? https : http;

            const bodyData = body !== null && body !== undefined ? JSON.stringify(body) : undefined;

            const options: http.RequestOptions = {
                method,
                hostname: url.hostname,
                port: url.port || (isHttps ? 443 : 80),
                path: url.pathname + url.search,
                timeout: this.timeoutMs,
                headers: {
                    'Authorization': `Bearer ${this.apiKey}`,
                    'Accept': 'application/json',
                    ...(bodyData ? { 'Content-Type': 'application/json', 'Content-Length': Buffer.byteLength(bodyData) } : {})
                }
            };

            const req = lib.request(options, (res) => {
                let data = '';

                res.on('data', (chunk) => {
                    data += chunk;
                });

                res.on('end', () => {
                    const statusCode = res.statusCode || 0;

                    if (statusCode < 200 || statusCode >= 300) {
                        let errorMessage = res.statusMessage || 'Unknown error';
                        let errorDetails: string | undefined;

                        if (data) {
                            try {
                                const errorResponse: ErrorResponse = JSON.parse(data);
                                errorMessage = errorResponse.Message || errorMessage;
                                errorDetails = errorResponse.Description;
                            } catch {
                                // Failed to parse error response
                            }
                        }

                        reject(new NetLedgerApiError(statusCode, errorMessage, errorDetails));
                        return;
                    }

                    if (!data || data.trim() === '') {
                        resolve({ StatusCode: statusCode } as ApiResponse<T>);
                        return;
                    }

                    try {
                        const parsedData = JSON.parse(data);
                        // Server sends data directly, wrap it in ApiResponse format
                        const response: ApiResponse<T> = {
                            StatusCode: statusCode,
                            Data: parsedData as T
                        };
                        resolve(response);
                    } catch (err) {
                        reject(new NetLedgerApiError(statusCode, 'Failed to parse response'));
                    }
                });
            });

            req.on('error', (err) => {
                reject(new NetLedgerConnectionError('Failed to connect to the server', err));
            });

            req.on('timeout', () => {
                req.destroy();
                reject(new NetLedgerConnectionError('Request timed out'));
            });

            if (bodyData) {
                req.write(bodyData);
            }
            req.end();
        });
    }
}
